src/procedures/proc_bfs.c

```diff
@@ -0,0 +1,222 @@
+	 * zero (unlimited), increment it by 1 to make level 1 indicate the source's direct neighbors. */
+	if(max_level > 0) max_level++;
+	const char *reltype = SIValue_IsNull(args[2]) ? NULL : args[2].stringval;
+	bool collect_edges = args[3].longval;
+
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	// Get edge matrix and transpose matrix, if available.
+	GrB_Matrix R;
+	GrB_Matrix TR;
+	if(reltype == NULL) {
+		R = Graph_GetAdjacencyMatrix(gc->g);
+		TR = Graph_GetTransposedAdjacencyMatrix(gc->g);
+	} else {
+		Schema *s = GraphContext_GetSchema(gc, reltype, SCHEMA_EDGE);
+		if(!s) return PROCEDURE_OK; // Failed to find schema, first step will return NULL.
+		pdata->reltype_id = s->id;
+		R = Graph_GetRelationMatrix(gc->g, s->id);
+		if(Config_MaintainTranspose()) TR = Graph_GetTransposedRelationMatrix(gc->g, s->id);
+		else TR = GrB_NULL;
+	}
+
+	GrB_Vector V;  // Vector of results
+	GrB_Vector PI = NULL; // Vector backtracking results to their parents.
+	GrB_Vector *PI_ptr = &PI;
+	if(!collect_edges) {
+		/* If we're not collecting edges, pass a NULL parent pointer so that the algorithm will
+		 * not perform unnecessary work. */
+		PI_ptr = GrB_NULL;
+		pdata->edges_output_idx = -1;
+	}
+	assert(LAGraph_bfs_both(&V, PI_ptr, R, TR, source_id, max_level, false) == GrB_SUCCESS);
+	/* Remove all values with a level less than or equal to 1.
+	 * Values of 0 are not connected to the source, and values of 1 are the source. */
+	GxB_Scalar thunk;
+	GxB_Scalar_new(&thunk, GrB_UINT64);
+	GxB_Scalar_setElement_UINT64(thunk, 1);
+	GxB_Vector_select(V, GrB_NULL, GrB_NULL, GxB_GT_THUNK, V, thunk, GrB_NULL);
+	GxB_Scalar_free(&thunk);
+
+	// Get number of entries.
+	GrB_Index nvals;
+	GrB_Vector_nvals(&nvals, V);
+	pdata->n = nvals;
+
+	// Retrieve all tuples representing connected nodes.
+	GrB_Index *node_ids = rm_malloc(nvals * sizeof(GrB_Index));
+	GrB_Vector_extractTuples_INT64(node_ids, GrB_NULL, &nvals, V);
+	pdata->node_ids = node_ids;
+	pdata->parents = PI;
+
+	GrB_Vector_free(&V);
+	return PROCEDURE_OK;
+}
+
+static SIValue *Proc_BFS_Step(ProcedureCtx *ctx) {
+	assert(ctx->privateData);
+
+	BFSContext *pdata = (BFSContext *)ctx->privateData;
+
+	// Return NULL if this source has already been mapped or there are no connected nodes.
+	if(pdata->depleted || pdata->n == 0) return NULL;
+
+	// Build arrays for the outputs the user has requested.
+	if(pdata->nodes_output_idx >= 0) pdata->output[pdata->nodes_output_idx] = SI_Array(pdata->n);
+	if(pdata->edges_output_idx >= 0) pdata->output[pdata->edges_output_idx] = SI_Array(pdata->n);
+
+	for(int i = 0; i < pdata->n; i ++) {
+		// Get the reached node.
+		NodeID node_id = pdata->node_ids[i];
+
+		if(pdata->nodes_output_idx >= 0) {
+			// Append each reachable node to the nodes output array.
+			Node n = GE_NEW_NODE();
+			Graph_GetNode(pdata->g, node_id, &n);
+			SIArray_Append(&pdata->output[pdata->nodes_output_idx], SI_Node(&n));
```

Comment:
Wouldn't multiple iterations of loop at line 110 would cause an overwrite of `n` ?
as `Node n` is on the stack and we're passing a pointer to it as a new entry of `pdata->output`

Reply:
`SIArray_Append` clones its input value, so each iteration of this loop is incapable of overwriting the values that have actually been placed in the array.